modules/apps/transfer/keeper/msg_server.go

```diff
@@ -32,12 +32,15 @@ func (k Keeper) Transfer(goCtx context.Context, msg *types.MsgTransfer) (*types.
 		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to send funds", sender)
 	}
 
+	// Use a cached context to prevent accidental state changes when seting the total amount in escrow
+	cacheCtx, writeFn := ctx.CacheContext()
```

Comment:
I would like @colin-axner or @damiannolan to check that this is the best way to achieve this.

Reply:
Looks unnecessary. The context is [already cached by baseapp](https://github.com/cosmos/cosmos-sdk/blob/main/baseapp/baseapp.go#L729), so if an error is returned by `sendTransfer` it will bubble up and state will not be commited by `writeCache()` at the sdk level.

Reply:
All right, thanks, @damiannolan. I added it because when running [this test](https://github.com/cosmos/ibc-go/blob/0903989307e2fe0b1d272eb3c0fd8ee87b42e731/modules/apps/transfer/keeper/relay_test.go#L102) the state seemed to not be reverted... I will check once more, though.

Reply:
In the testcase its calling the keeper function directly, so the msg is never being routed through baseapp. 
You could use `chain.SendMsgs(msgTransfer)` to route it through baseapp, but I think that'd require a good bit of refactoring to the testcases